From e9928c94ce1385ec72028336417f19f30ac38ac0 Mon Sep 17 00:00:00 2001
From: Nezih Yigitbasi <nyigitbasi@netflix.com>
Date: Fri, 25 Mar 2016 12:19:39 -0700
Subject: [PATCH] PARQUET-571: Fix potential leak in ParquetFileReader.close()

If an exception occurs when closing the input stream `f`, the codecs
will not be released. This may cause native memory leaks for some codecs. \cc @rdblue

Author: Nezih Yigitbasi <nyigitbasi@netflix.com>

Closes #338 from nezihyigitbasi/leak-fix and squashes the following commits:

fcc5528 [Nezih Yigitbasi] Fix potential leak in close()
---
 .../apache/parquet/hadoop/ParquetFileReader.java    | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
index 3d7b499a81..7ac1706c7b 100644
--- a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
+++ b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java
@@ -776,11 +776,14 @@ private static DictionaryPage readCompressedDictionary(
 
   @Override
   public void close() throws IOException {
-    if (f != null) {
-      f.close();
-    }
-    if (codecFactory != null) {
-      codecFactory.release();
+    try {
+      if (f != null) {
+        f.close();
+      }
+    } finally {
+      if (codecFactory != null) {
+        codecFactory.release();
+      }
     }
   }
 
